akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala

```diff
@@ -191,52 +192,79 @@ object Framing {
 
-    override def onUpstreamFinish(ctx: Context[ByteString]): TerminationDirective =
-      if (buffer.nonEmpty) ctx.absorbTermination()
-      else ctx.finish()
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      private var buffer = ByteString.empty
+      private var frameSize = Int.MaxValue
 
-    private def doParse(ctx: Context[ByteString]): SyncDirective = {
-      def emitFrame(ctx: Context[ByteString]): SyncDirective = {
-        val parsedFrame = buffer.take(frameSize).compact
+      /**
+       * push, and reset frameSize and buffer
+       *
+       */
+      private def pushFrame = {
+        val emit = buffer.take(frameSize).compact
         buffer = buffer.drop(frameSize)
         frameSize = Int.MaxValue
-        if (ctx.isFinishing && buffer.isEmpty) ctx.pushAndFinish(parsedFrame)
-        else ctx.push(parsedFrame)
+        push(out, emit)
       }
 
-      val bufSize = buffer.size
-      if (bufSize >= frameSize) emitFrame(ctx)
-      else if (bufSize >= minimumChunkSize) {
-        val parsedLength = intDecoder(buffer.iterator.drop(lengthFieldOffset), lengthFieldLength)
-        frameSize = parsedLength + minimumChunkSize
-        if (frameSize > maximumFrameLength)
-          ctx.fail(new FramingException(s"Maximum allowed frame size is $maximumFrameLength but decoded frame header reported size $frameSize"))
-        else if (bufSize >= frameSize) emitFrame(ctx)
-        else tryPull(ctx)
-      } else tryPull(ctx)
-    }
+      /**
+       * try to push downstream, returns true if push success
+       *
+       * @return
+       */
+      private def tryPushFrame: Boolean = {
+        var result = false
+        if (buffer.size >= frameSize) {
+          pushFrame
+          result = true
+        } else if (buffer.size >= minimumChunkSize) {
+          val parsedLength = intDecoder(buffer.iterator.drop(lengthFieldOffset), lengthFieldLength)
+          frameSize = parsedLength + minimumChunkSize
+          if (frameSize > maximumFrameLength) {
+            failStage(new FramingException(s"Maximum allowed frame size is $maximumFrameLength but decoded frame header reported size $frameSize"))
+            result = false
+          } else if (buffer.size >= frameSize) {
+            pushFrame
+            result = true
+          }
+        }
+        result
+      }
 
-    override def postStop(): Unit = buffer = null
+      override def onPush(): Unit = {
+        buffer ++= grab(in)
+        if (!tryPushFrame) {
+          pull(in)
+        }
+      }
+
+      override def onPull(): Unit = {
+        if (!tryPushFrame) {
+          pull(in)
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (buffer.size > 0) {
+          failStage(new FramingException("Stream finished but there was a truncated final frame in the buffer"))
```

Comment:
@johanandren I think it's conformed to the original one, see [this](https://github.com/akka/akka/pull/20345/files#diff-263c698cb6645f273cf54e2961c3e8a1L233), and there is also a unit test for this, see [FramingSpec](https://github.com/akka/akka/blob/master/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FramingSpec.scala#L104).